### PR TITLE
Split out testing extras

### PR DESCRIPTION
**Pull request by bmw**
I think this PR makes how we declare extra dependencies ever so slightly saner. I want this for https://github.com/certbot/certbot/issues/8787 as there are dependency conflicts between things like `poetry` and some of our dependencies in the oldest tests.

I considered thinking of our extras as part of our stable API, but I don't think we should due to our note at https://certbot.eff.org/docs/compatibility.html and me not finding a good way to deprecate an extra.

I initially started to make more fine grained extras like `coverage`, `lint`, and `mypy`, however, I think it's a little difficult with the way our `tox.ini` file is structured. Instead, I just made one `test` extra which will install all of our dependencies used for automated testing.

`acme` still needs a `test` extra so it works with `install_and_test.py` used in `tox.ini` where it is tested before Certbot is installed.

The dependency on `asteroid` seems unnecessary since https://github.com/certbot/certbot/pull/7551. We do not seem to import the library anywhere in our code. Before that PR, it was used to pin the version of `asteroid` used by `pylint`.

I also made the `all` extra for convenience since we want to install everything in a couple places and it simplifies things if we add more extras to Certbot in the future.

I left `tox` out of "test" extras because with the exception of the modified test farm test, it was never needed.

### CODE REVIEW

File: tools/pinning/pyproject.toml [hunk 1]
```diff
@@ -30,8 +30,8 @@ certbot-dns-route53 = {path = "../../certbot-dns-route53", extras = ["docs"]}
 certbot-dns-sakuracloud = {path = "../../certbot-dns-sakuracloud", extras = ["docs"]}
 certbot-nginx = {path = "../../certbot-nginx", extras = ["docs"]}
 certbot-apache = {path = "../../certbot-apache", extras = ["dev"]}
-certbot = {path = "../../certbot", extras = ["dev", "docs"]}
-acme = {path = "../../acme", extras = ["dev", "docs"]}
+certbot = {path = "../../certbot", extras = ["all"]}
+acme = {path = "../../acme", extras = ["test"]}
```
Comment (ohemorange) [inline tools/pinning/pyproject.toml:34]: Why don't we need `docs` here, given that all the other modules have it?

**Review by ohemorange — COMMENTED** [review]
I notice we pin apache dev as well, and docs for many others. Is the idea to do the others later or are they just not as complicated and so don't need it?

File: tools/pinning/pyproject.toml (hunk 1 above)
Comment (bmw) [reply tools/pinning/pyproject.toml:34]: This was a mistake. I didn't mean to drop docs here. Nice catch.

**Comment by bmw** [comment]
> I notice we pin apache dev as well, and docs for many others. Is the idea to do the others later or are they just not as complicated and so don't need it?

It's that we don't define dev or docs extras for the other packages:
```
$ grep -e "'dev'" -e '"dev"' */setup.py
certbot-apache/setup.py:        'dev': dev_extras,
certbot/setup.py:        'dev': dev_extras,
$ grep -e "'docs'" -e '"docs"' */setup.py
acme/setup.py:        'docs': docs_extras,
certbot-dns-cloudflare/setup.py:        'docs': docs_extras,
certbot-dns-cloudxns/setup.py:        'docs': docs_extras,
certbot-dns-digitalocean/setup.py:        'docs': docs_extras,
certbot-dns-dnsimple/setup.py:        'docs': docs_extras,
certbot-dns-dnsmadeeasy/setup.py:        'docs': docs_extras,
certbot-dns-gehirn/setup.py:        'docs': docs_extras,
certbot-dns-google/setup.py:        'docs': docs_extras,
certbot-dns-linode/setup.py:        'docs': docs_extras,
certbot-dns-luadns/setup.py:        'docs': docs_extras,
certbot-dns-nsone/setup.py:        'docs': docs_extras,
certbot-dns-ovh/setup.py:        'docs': docs_extras,
certbot-dns-rfc2136/setup.py:        'docs': docs_extras,
certbot-dns-route53/setup.py:        'docs': docs_extras,
certbot-dns-sakuracloud/setup.py:        'docs': docs_extras,
certbot/setup.py:    packages=find_packages(exclude=['docs', 'examples', 'tests', 'venv']),
certbot/setup.py:        'docs': docs_extras,
```
It looks like this file was depending on docs extras that didn't exist though. I'm disappointed that poetry doesn't error about that, but I fixed up the file so we stop doing that.

**Comment by bmw** [comment]
Ugh. I regenerated the pinnings since I modified the extras we depend on just in case, but doing so updated mypy to the same version that broke josepy last night and it's going to break the Certbot tests too.

Any preference on how I resolve this in this PR? Some options are:

1. Revert changes to pinnings file in the most recent commit since they weren't necessary.
2. Pin back mypy in this PR with a comment about new versions breaking and make an issue to fix it.
3. Fix mypy in this PR or block this PR on fixing mypy.

**Comment by ohemorange** [comment]
I'd prefer 2, then 3, then 1, just because it'd make me more comfortable to make sure it runs correctly.

**Comment by bmw** [comment]
Here's option 2. I'll either open a PR fixing it or open an issue about the problem later today.

**Review by ohemorange — APPROVED** [review]
LGTM!

**Pull request merged**